## Code - Method-1 :: Using Anova Without RBD(Randomized Block Approach)

In [5]:
import pandas as pd
import numpy as np
import scipy
import statsmodels.api  as sm        ## for statsmodels.api.stats.anova_lm
from statsmodels.formula.api import ols

In [7]:
path = '/home/ashish/projects/RandomizeBlockDesign_FabricStrength/rbd2.xlsx'
data_frame = pd.read_excel(path)
data_frame

chem1  chem2  chem3  chem4
0    1.3    2.2    1.8    3.9
1    1.6    2.4    1.7    4.4
2    0.5    0.4    0.6    2.0
3    1.2    2.0    1.5    4.1
4    1.1    1.8    1.3    3.4

In [13]:
id_vars = ['index']
value_vars = ['chem1','chem2','chem3','chem4']
melted_data = pd.melt(data_frame.reset_index(),id_vars,value_vars)
melted_data.columns = ['indexing','treatments','values']
melted_data

indexing treatments  values
0          0      chem1     1.3
1          1      chem1     1.6
2          2      chem1     0.5
3          3      chem1     1.2
4          4      chem1     1.1
5          0      chem2     2.2
6          1      chem2     2.4
7          2      chem2     0.4
8          3      chem2     2.0
9          4      chem2     1.8
10         0      chem3     1.8
11         1      chem3     1.7
12         2      chem3     0.6
13         3      chem3     1.5
14         4      chem3     1.3
15         0      chem4     3.9
16         1      chem4     4.4
17         2      chem4     2.0
18         3      chem4     4.1
19         4      chem4     3.4

In [14]:
model = ols('values ~ C(treatments)',data=melted_data).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 values   R-squared:                       0.702
Model:                            OLS   Adj. R-squared:                  0.647
Method:                 Least Squares   F-statistic:                     12.59
Date:                Sat, 25 Dec 2021   Prob (F-statistic):           0.000176
Time:                        19:16:11   Log-Likelihood:                -18.761
No. Observations:                  20   AIC:                             45.52
Df Residuals:                      16   BIC:                             49.50
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                  1.1400      0.309      3.688      0.002       0.485       1.795
C(treatments)[T.chem2]     0.6200      0.437      1.418      0.175      -0.307       1.547
C(treatments)[T.chem3]     0.2400      0.437      0.549      0.591      -0.687       1.167
C(treatments)[T.chem4]     2.4200      0.437      5.536      0.000       1.493       3.347
==============================================================================
Omnibus:                        7.528   Durbin-Watson:                   2.874
Prob(Omnibus):                  0.023   Jarque-Bera (JB):                5.231
Skew:                          -1.205   Prob(JB):                       0.0731
Kurtosis:                       3.688   Cond. No.                         4.79
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [15]:
anova_table = sm.stats.anova_lm(model,typ=1)
anova_table

df  sum_sq   mean_sq          F    PR(>F)
C(treatments)   3.0  18.044  6.014667  12.589569  0.000176
Residual       16.0   7.644  0.477750        NaN       NaN

# So here since 
                p-value = 0.000176 < alpha=0.01
                SO WE REJECT THE NULL HYPOTHESIS(all means are equal).

# -----------------------------------------------------------------------

# Method-2:: Using the Randomized Block Approach.

In [28]:
import pandas as pd
import numpy as np
import scipy
import statsmodels as sm
from statsmodels.formula.api import ols

In [29]:
path_2 = '/home/ashish/projects/RandomizeBlockDesign_FabricStrength/rbd2.xlsx'
data_frame_2 = pd.read_excel(path_2)
data_frame_2

chem1  chem2  chem3  chem4
0    1.3    2.2    1.8    3.9
1    1.6    2.4    1.7    4.4
2    0.5    0.4    0.6    2.0
3    1.2    2.0    1.5    4.1
4    1.1    1.8    1.3    3.4

In [30]:
id_vars = ['index']
value_vars = ['chem1','chem2','chem3','chem4']
melted_data_2 = pd.melt(data_frame_2.reset_index(),id_vars,value_vars)
melted_data_2.columns = ['Fabric','Chemical','values']
melted_data_2

Fabric Chemical  values
0        0    chem1     1.3
1        1    chem1     1.6
2        2    chem1     0.5
3        3    chem1     1.2
4        4    chem1     1.1
5        0    chem2     2.2
6        1    chem2     2.4
7        2    chem2     0.4
8        3    chem2     2.0
9        4    chem2     1.8
10       0    chem3     1.8
11       1    chem3     1.7
12       2    chem3     0.6
13       3    chem3     1.5
14       4    chem3     1.3
15       0    chem4     3.9
16       1    chem4     4.4
17       2    chem4     2.0
18       3    chem4     4.1
19       4    chem4     3.4

In [31]:
model = ols('values ~ C(Fabric) + C(Chemical)',data = melted_data_2).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 values   R-squared:                       0.963
Model:                            OLS   Adj. R-squared:                  0.941
Method:                 Least Squares   F-statistic:                     44.59
Date:                Sat, 25 Dec 2021   Prob (F-statistic):           1.18e-07
Time:                        19:32:18   Log-Likelihood:                 2.0810
No. Observations:                  20   AIC:                             11.84
Df Residuals:                      12   BIC:                             19.80
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
Intercept                1.4800      0.178      8.313      0.000       1.092       1.868
C(Fabric)[T.1]           0.2250      0.199      1.130      0.280      -0.209       0.659
C(Fabric)[T.2]          -1.4250      0.199     -7.159      0.000      -1.859      -0.991
C(Fabric)[T.3]          -0.1000      0.199     -0.502      0.625      -0.534       0.334
C(Fabric)[T.4]          -0.4000      0.199     -2.009      0.068      -0.834       0.034
C(Chemical)[T.chem2]     0.6200      0.178      3.482      0.005       0.232       1.008
C(Chemical)[T.chem3]     0.2400      0.178      1.348      0.203      -0.148       0.628
C(Chemical)[T.chem4]     2.4200      0.178     13.592      0.000       2.032       2.808
==============================================================================
Omnibus:                        0.096   Durbin-Watson:                   3.104
Prob(Omnibus):                  0.953   Jarque-Bera (JB):                0.055
Skew:                           0.043   Prob(JB):                        0.973
Kurtosis:                       2.758   Cond. No.                         6.42
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [34]:
anova_table = sm.api.stats.anova_lm(model,typ=1)
anova_table

df  sum_sq   mean_sq          F        PR(>F)
C(Fabric)     4.0   6.693  1.673250  21.113565  2.318913e-05
C(Chemical)   3.0  18.044  6.014667  75.894848  4.518310e-08
Residual     12.0   0.951  0.079250        NaN           NaN

# HERE ALSO WE HAVE REJECTED THE NULL HYPOTHESIS ---- using the blocking approach.

# ******* NOTE : WE HAVE REDUCED THE MEAN SQUARED ERROR AND SUM SQUARED ERROR BY AN SIGNIFICANT AMOUNT USING THE BLOCKING OF NUUISANCE/EXTRANEOUS FACTORS. 